### PR TITLE
Redesign CPU in VM details

### PR DESCRIPTION
**Pull request by skobyda**
This is a preparation PR for a redesign of Details page: https://github.com/cockpit-project/cockpit-machines/issues/1027

This PR only redesign CPU-related things:
- It merges CPU model dialog and vCPU dialog into a single dialog
- Get's rid of column layout for vCPU, using a simple vertical form

It also does few small design improvements suggesteed by @garrett :
- Uses PF's NumberInput component for vCPU count and vCPU maximum
- Moves vCPU maximum before vCPU count
- Changes dialog size from medium to small
- Uses HelpIcon instead of InfoAltIcon
- Makes width of selectors smaller, since they were unecessary wide

There is also issue with HelpIcon being too low, but that's not issue of only CPU dialog, but everywhere in cockpit. I'm trying to fix it at https://github.com/cockpit-project/cockpit/pull/18640

### Before
Screenshot from 2023-04-14 13-53-54
### After
Screenshot from 2023-04-14 13-47-22
### Before
Screenshot from 2023-04-14 13-54-08
Screenshot from 2023-04-14 13-54-00
### After
Screenshot from 2023-04-14 13-47-00

### CODE REVIEW

**Comment by skobyda** [comment]
Btw this design was discussed yesterday with @garrett where we quasi-mocked it, so I believe the design will not change anymore (or only marginally), so once I fix the test, I think we can immediatelly start with code-review.

File: src/components/vm/overview/cpuModal.jsx [hunk 1]
```diff
@@ -0,0 +1,326 @@
+import React, { useState } from 'react';
```
Comment (KKoukiou) [inline src/components/vm/overview/cpuModal.jsx:1]: Missing licence

File: src/components/vm/overview/cpuModal.jsx [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+            return;
+        }
+
+        const newValue = clamp(value, max, 1);
+        setCount(parseInt(newValue));
+    }
+
+    function onSocketChange (value) {
+        const state = { sockets, cores };
+        state.sockets = parseInt(value);
```
Comment (KKoukiou) [inline src/components/vm/overview/cpuModal.jsx:112]: This is an anti-practice - this is changing the component state without the set* function.

File: src/components/vm/overview/cpuModal.jsx [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        // Get divisors of Max VCPU number divided by number of sockets
+        const divs = dividers(max / state.sockets);
+
+        // If current cores value is not in divisors array, then change it to max divisor
+        if (divs.indexOf(cores) === -1) {
+            state.cores = divs[divs.length - 1];
+        }
+
+        // Likewise: Max value = Sockets * Cores * Threads. Sockets = Max value / ( Threads * Cores )
+        state.threads = (max / (state.sockets * state.cores));
```
Comment (KKoukiou) [inline src/components/vm/overview/cpuModal.jsx:123]: Same here.

**Review by KKoukiou — COMMENTED** [review]
You have a few state updates without using the react's state modifier functions. Otherwsie looks good from first pass.

File: src/components/vm/overview/cpuModal.jsx (hunk 2 above)
Comment (skobyda) [reply src/components/vm/overview/cpuModal.jsx:112]: But this is not the `state` itself, it's just a local variable, local object `const state = { sockets, cores };`, So changing `state.socket = something` doesn't change react state, only the local variable.
But I admit the name of variable is confusing... perhaps let's call it stateDelta, or newState?

**Comment by martinpitt** [comment]
RPM build failure in COPR/TF is fixed, but as our own integration tests fail as well, this needs another push anyway.

**Comment by skobyda** [comment]
@garrett mind giving it review?

**Review by garrett — CHANGES_REQUESTED** [review]
Oh, I wasn't aware this was ready. The host "mode" is still at the top, which makes the order on label on the card and in the dialog inconsistent. They should be consistent, so the mode should probably go at the bottom of the modal, as we discussed in our call.

image

When I tried to change to passthrough, I got this error message and the spinner never completes. It's possibly unrelated to this PR, but it might also be a side effect of this PR, as we're changing both at the same time now:

image

**Comment by skobyda** [comment]
> When I tried to change to passthrough, I got this error message and the spinner never completes. It's possibly unrelated to this PR, but it might also be a side effect of this PR, as we're changing both at the same time now:

This is unrelated to this redesign, you can achieve the same with the old design:
Screenshot from 2023-04-20 11-39-44

But I'm more interested, how did you manage to get into this situation that you have 2 vCPUs buy only 1 socket, 1 core per socket and 1 thread per core. You should not be able to set that from our UI. But you can set it like that through virsh or virt-manager, since they still support the "old way (before qemu 2.6"), where maxVCPU didn't have to match the cpu topology, maybe that's how you got that configuration?

**Comment by skobyda** [comment]
> When it's host passthrough, are you able to specify all of the values of the topology? If not, we should hide what you can't change (and make sure the values work with it).


You are still able to specify all fields, no matter what mode you choose. 

> However, if the mode does influence the topology, then we probably want to keep mode at the top and change the label in the overview card to something like: "**CPU** Host mode, 2vCPUs" and "**CPU** Passthrough mode, 2vCPUs" right?

No, mode doesn't influence topology
 
Updated the design:
Screenshot from 2023-04-20 11-46-40

**Comment by skobyda** [comment]
@KKoukiou @martinpitt mind giving this a code-review?

**Review by garrett — DISMISSED** [review]
Thanks for flipping the dropdown to the bottom! This is good from a design viewpoint now.

We _could_ redesign it further later, but for a simple merge with a simple design, this now works for me! :+1: Thanks for taking the initiative at merging it and implementing this, @skobyda!

**Comment by skobyda** [comment]
Fixed pixel tests, now it should be all green

**Pull request merged**